modules/library/main/src/main/java/org/geotools/data/store/FilteringFeatureIterator.java

```diff
@@ -67,6 +67,9 @@ public boolean hasNext() {
     }
 
     public F next() throws NoSuchElementException {
+        if(next==null&&this.hasNext()==false) {
+            throw new NoSuchElementException();
```

Comment:
minor suggestions:
please, check the formatting (need some whitespaces). 
I would also personally use !this.hasNext() instead of explicit check with false.

if (next == null && !this.hasNext()) {